src/main/scala/viper/gobra/reporting/StatsCollector.scala

```diff
@@ -0,0 +1,357 @@
+        .filter(_.key != this.key).toSet
+    }
+
+    def key: String = gobraMemberKey(pkgDir, pkg, memberName, args)
+  }
+
+  case class ViperMemberEntry(memberName: String,
+                              taskName: String,
+                              time: Time,
+                              nodeType: String,
+                              dependencies: Set[String],
+                              success: Boolean,
+                              cached: Boolean,
+                              fromImport: Boolean,
+                              hasBody: Boolean) {
+
+    def asJson(p: String = ""): String = {
+      s"""$p{
+         |$p  "name": "${this.memberName}",
+         |$p  "taskName": "${this.taskName}",
+         |$p  "time": ${this.time},
+         |$p  "nodeType": "${this.nodeType}",
+         |$p  "success": ${this.success},
+         |$p  "cached": ${this.cached},
+         |$p  "fromImport": ${this.fromImport},
+         |$p  "hasBody": ${this.hasBody}
+         |$p}""".stripMargin
+    }
+  }
+
+  case class GobraMemberInfo(pkgDir: String,
+                             pkg: String,
+                             memberName: String,
+                             args: String,
+                             isTrusted: Boolean,
+                             isAbstract: Boolean,
+                             isImported: Boolean,
+                             isBuiltIn: Boolean)
+
+  override val name: String = "StatsCollector"
+
+  override def report(msg: GobraMessage): Unit = {
+    def handleEntityMessage(taskName: String, viperMember: Member, info: Source.Verifier.Info, time: Time, cached: Boolean, success: Boolean): Unit = {
+      Violation.violation(typeInfos.contains(taskName), "No type info available for stats reporter")
+      getMemberInformation(info.pnode, typeInfos(taskName)) match {
+        case GobraMemberInfo(pkgDir, pkg, memberName, args, isTrusted, isAbstract, isImported, false) =>
+          addResult(
+            pkgDir,
+            pkg,
+            memberName,
+            info.pnode.getClass.getSimpleName,
+            args,
+            ViperMemberEntry(
+              viperMemberName(viperMember),
+              taskName,
+              time,
+              viperMember.getClass.getSimpleName,
+              getViperDependencies(viperMember),
+              success,
+              cached,
+              isImported,
+              viperMemberHasBody(viperMember)
+            ),
+            isTrusted,
+            isAbstract)
+        case _ =>
+      }
+    }
+
+    msg match {
+      // Capture typeInfo once it's available
+      case TypeInfoMessage(typeInfo, taskName) => this.synchronized({ typeInfos = typeInfos + (taskName -> typeInfo)})
+      // Free up unneeded space, once a task is finished
+      case VerificationTaskFinishedMessage(taskName) => this.synchronized({ typeInfos = typeInfos - taskName})
```

Comment:
In what contexts could we have a concurrent access to the map? If we really need it, why not use a ConcurrentHashMap for this map? That seems like a less brittle alternative to having synchronized blocks which we can easily forget about when the code is updated